source/fab/artifact.py

```diff
@@ -0,0 +1,113 @@
+##############################################################################
+# (c) Crown copyright Met Office. All rights reserved.
+# For further details please refer to the file COPYRIGHT
+# which you should have received as part of this distribution
+##############################################################################
+
+from pathlib import Path
+from abc import ABC
+from typing import Type, List, Optional
+from zlib import adler32
+
+from fab.reader import FileTextReader
+
+
+# Classes representing possible states of an Artifact
+class State(ABC):
+    pass
+
+
+class New(State):
+    pass
+
+
+class Seen(State):
+    pass
+
+
+class Raw(State):
+    pass
+
+
+class Analysed(State):
+    pass
+
+
+class Compiled(State):
+    pass
+
+
+class Linked(State):
+    pass
+
+
+# Classes representing possible filetypes
+class FileType(ABC):
+    pass
+
+
+class Unknown(FileType):
+    pass
+
+
+class FortranSource(FileType):
+    pass
+
+
+class BinaryObject(FileType):
+    pass
+
+
+class Executable(FileType):
+    pass
+
+
+class Artifact(object):
+    def __init__(self,
+                 location: Path,
+                 filetype: Type[FileType],
+                 state: Type[State]) -> None:
+
+        self._location = location
+        self._filetype = filetype
+        self._state = state
+        self._defines: List[str] = []
+        self._depends_on: List[str] = []
+        self._hash: Optional[int] = None
+
+    @property
+    def location(self) -> Path:
+        return self._location
+
+    @property
+    def filetype(self) -> Type[FileType]:
+        return self._filetype
+
+    @property
+    def state(self) -> Type[State]:
+        return self._state
+
+    @property
+    def defines(self) -> List[str]:
+        return self._defines
+
+    @property
+    def depends_on(self) -> List[str]:
+        return self._depends_on
+
+    @property
+    def hash(self) -> int:
+        # If this is the first access of the property calculate the hash
+        # and cache it for later accesses
+        if self._hash is None:
+            self._hash = 1
+            reader = FileTextReader(self.location)
+            for line in reader.line_by_line():
+                self._hash = adler32(bytes(line, encoding='utf-8'), self._hash)
+        return self._hash
```

Comment:
On demand hashing - nice.

Reply:
Hopefully it comes in useful - it's not totally clear to me yet exactly which files we will require hashes of, but now they'll just be available where needed